multiprocess_safe_cache

juj opened the pull request:
Implement a file-based lock mechanism to access the Emscripten cache. Closes #3850. Also fix an issue from earlier commit where at least on Windows, calling pool.terminate / pool.join might not succeed, but instead throws an "Access is denied" exception.

kripken commented on tools/cache.py:
```diff
@@ -17,15 +16,27 @@ def __init__(self, dirname=None, debug=False, use_subdir=True):
     self.dirname = dirname
     self.debug = debug
 
+    def reverse_chop(thestring, ending):
```
the name seems surprising? perhaps "try_remove_ending"?

juj replied:
Ok, that's a good name.

kripken commented on tools/duplicate_function_eliminator.py:
```diff
@@ -247,8 +247,11 @@ def write_chunk(chunk, i):
+      try:
+        pool.terminate()
+        pool.join()
+      except Exception, e:
```
what is caught here? a comment would help

juj replied:
Documented this and the instance below better.

kripken replied:
I still don't get it. What is being caught - why is an error now possible that was not before?

kripken replied:
To be clearer: I still don't get why the change is needed. Why did this PR make those windows errors suddenly happen?

juj replied:
Oh sorry, my bad in explaining here. This PR did not cause those errors, but those have been always occurring with low frequency, on certain machines. The new Windows bot runs into those quite often, see the latest Windows bot run for example: http://clb.demon.fi:8112/builders/win-incoming/builds/1284/steps/default%20suite/logs/stdio

and search for "Access is denied". One such dump:

    test_dlfcn_unique_sig (test_core.asm3) ... Traceback (most recent call last):
      File "C:\slaves\emslave\buildslave\win-incoming\emsdk\emscripten\incoming\emcc", line 13, in <module>
        emcc.run()
      File "C:\slaves\emslave\buildslave\win-incoming\emsdk\emscripten\incoming\emcc.py", line 1784, in run
        JSOptimizer.flush()
      File "C:\slaves\emslave\buildslave\win-incoming\emsdk\emscripten\incoming\emcc.py", line 1681, in flush
        run_passes(chunks[0], title, just_split=False, just_concat=False)
      File "C:\slaves\emslave\buildslave\win-incoming\emsdk\emscripten\incoming\emcc.py", line 1654, in run_passes
        final = shared.Building.js_optimizer(final, passes, debug_level >= 4, JSOptimizer.extra_info, just_split=just_split, just_concat=just_concat)
      File "C:\slaves\emslave\buildslave\win-incoming\emsdk\emscripten\incoming\tools\shared.py", line 1748, in js_optimizer
        ret = js_optimizer.run(filename, passes, NODE_JS, debug, extra_info, just_split, just_concat)
      File "C:\slaves\emslave\buildslave\win-incoming\emsdk\emscripten\incoming\tools\js_optimizer.py", line 544, in run
        return temp_files.run_and_clean(lambda: run_on_js(filename, passes, js_engine, source_map, extra_info, just_split, just_concat))
      File "C:\slaves\emslave\buildslave\win-incoming\emsdk\emscripten\incoming\tools\tempfiles.py", line 64, in run_and_clean
        return func()
      File "C:\slaves\emslave\buildslave\win-incoming\emsdk\emscripten\incoming\tools\js_optimizer.py", line 544, in <lambda>
        return temp_files.run_and_clean(lambda: run_on_js(filename, passes, js_engine, source_map, extra_info, just_split, just_concat))
      File "C:\slaves\emslave\buildslave\win-incoming\emsdk\emscripten\incoming\tools\js_optimizer.py", line 447, in run_on_js
        pool.terminate()
      File "C:\slaves\emslave\buildslave\win-incoming\emsdk\python\2.7.5.3_64bit\lib\multiprocessing\pool.py", line 452, in terminate
        self._terminate()
      File "C:\slaves\emslave\buildslave\win-incoming\emsdk\python\2.7.5.3_64bit\lib\multiprocessing\util.py", line 201, in __call__
        res = self._callback(*self._args, **self._kwargs)
      File "C:\slaves\emslave\buildslave\win-incoming\emsdk\python\2.7.5.3_64bit\lib\multiprocessing\pool.py", line 500, in _terminate_pool
        p.terminate()
      File "C:\slaves\emslave\buildslave\win-incoming\emsdk\python\2.7.5.3_64bit\lib\multiprocessing\process.py", line 137, in terminate
        self._popen.terminate()
      File "C:\slaves\emslave\buildslave\win-incoming\emsdk\python\2.7.5.3_64bit\lib\multiprocessing\forking.py", line 312, in terminate
        _subprocess.TerminateProcess(int(self._handle), TERMINATE)
    WindowsError: [Error 5] Access is denied
    Error in atexit._run_exitfuncs:
    Traceback (most recent call last):
      File "C:\slaves\emslave\buildslave\win-incoming\emsdk\python\2.7.5.3_64bit\lib\atexit.py", line 24, in _run_exitfuncs
        func(*targs, **kargs)
      File "C:\slaves\emslave\buildslave\win-incoming\emsdk\python\2.7.5.3_64bit\lib\multiprocessing\util.py", line 315, in _exit_function
        p._popen.terminate()
      File "C:\slaves\emslave\buildslave\win-incoming\emsdk\python\2.7.5.3_64bit\lib\multiprocessing\forking.py", line 312, in terminate
        _subprocess.TerminateProcess(int(self._handle), TERMINATE)
    WindowsError: [Error 5] Access is denied
    Error in sys.exitfunc:
    Traceback (most recent call last):
      File "C:\slaves\emslave\buildslave\win-incoming\emsdk\python\2.7.5.3_64bit\lib\atexit.py", line 24, in _run_exitfuncs
        func(*targs, **kargs)
      File "C:\slaves\emslave\buildslave\win-incoming\emsdk\python\2.7.5.3_64bit\lib\multiprocessing\util.py", line 315, in _exit_function
        p._popen.terminate()
      File "C:\slaves\emslave\buildslave\win-incoming\emsdk\python\2.7.5.3_64bit\lib\multiprocessing\forking.py", line 312, in terminate
        _subprocess.TerminateProcess(int(self._handle), TERMINATE)
    WindowsError: [Error 5] Access is denied
    FAIL

The root cause is the same as what is causing bug #718, but in that bug, the discussion related to a different code site where it stemmed from.

I could have made the catch only grab `WindowsError`s instead of all errors, but seems like if we get any error in this cleanup, the only sensible thing we can do is to just go on and ignore.

kripken replied:
Cool thanks, now I see.

lgtm!

kripken commented on tools/js_optimizer.py:
```diff
@@ -444,8 +444,11 @@ def write_chunk(chunk, i):
+      try:
+        pool.terminate()
+        pool.join()
+      except Exception, e:
```
same questions as above

kripken commented:
Do all tests (including and especially `sanity`) pass?

We should also add a test for this, in `sanity` - opening two processes that try to build say libc at the same time. What would happen with this lock, one would block until the other is done?

Could we show a warning to the console when blocking on the lock, so that if a problem happens and the locker is stuck or slow, the user knows what's going on? Or maybe the stack trace from control-c would be enough, if there is something useful on the stack.

juj commented:
Seeing some issues in sanity (#4416), but not related to this pull request. Working on the test.

juj commented:
Added a warning print if accessing the lock takes more than 60 seconds. Also added a test `other.test_emcc_multiprocess_cache_access` to stress this. Ran through `other` and `asm2` suites, which pass.

kripken commented:
60 seconds seems like a lot. I think people would assume it hangs and kill it much sooner. However, in EMCC_DEBUG mode we'd have logging shown, if I read this PR right? So I guess that's ok.

But it seems dangerous to wait 60 seconds and stop. How about after 60 seconds we show a warning (not just in debug output), and continue waiting? So if it frees up later it still works. But the user has an error message telling them why the delay, and maybe also saying how to manually kill the lock if necessary.

juj commented:
In debug mode there is a print before attempting to take the lock. The warning is printed after having waited for 60 seconds, but the code doesn't stop at the 60 seconds mark, but keeps waiting by re-attempting to take the lock without any timeout set (line 46, `self.filelock.acquire()`). So I think the behavior already is like you describe in the last paragraph.

The trouble with choosing a timeout is that building libc and/or any other thing to the cache can easily take more than 60 seconds on a slower computer or a VM, so anything smaller at least doesn't sound safe.

Also, this lock will not change performance or throughput in any way, so with respect to the performance/timing characteristics the behavior should be identical. The only thing this will delay is the 2nd and subsequent processes will not start each to stomp on building libc and other things to the same cache at the same time.

The only thing that I could think of where performance can get degraded is if two independent Emscripten processes wanted to build two completely different things to the cache, say e.g. build 1 wanted to build SDL2, and build 2 wanted to build freetype, although I expect such circumstances would be extremely rare.

kripken commented:
Ok, sounds good.

My last concern is the question in the other thread, why we need to catch new windows errors now.

juj commented:
Thanks, merging. I'm hopeful that Windows would go green on the default run mode after this.

kripken commented:
Great. Btw, maybe worth posting to the mailing list to ask for people to test this?

juj commented:
Good idea, let me advertise a bit in there.
